hypervisor/arch/x86/timer.c

```diff
@@ -278,23 +278,30 @@ static uint64_t pit_calibrate_tsc(uint32_t cal_ms_arg)
 }
 
 /*
- * Determine TSC frequency via CPUID 0x15
+ * Determine TSC frequency via CPUID 0x15 and 0x16.
  */
 static uint64_t native_calibrate_tsc(void)
 {
+	uint64_t tsc_hz = 0;
+
 	if (boot_cpu_data.cpuid_level >= 0x15U) {
 		uint32_t eax_denominator, ebx_numerator, ecx_hz, reserved;
 
 		cpuid(0x15U, &eax_denominator, &ebx_numerator,
 			&ecx_hz, &reserved);
 
 		if (eax_denominator != 0U && ebx_numerator != 0U) {
-			return ((uint64_t) ecx_hz *
+			tsc_hz = ((uint64_t) ecx_hz *
 				ebx_numerator) / eax_denominator;
 		}
 	}
+	if ((tsc_hz == 0) && (boot_cpu_data.cpuid_level >= 0x16U)) {
```

Comment:
should we move this earlier than leaf 0x15? That will be slightly simple

Reply:
TSC based on 0x15 leaf is more precise than that based on 0x16 leaf.
So it will be better that it checks the 0x16 leaf after zero tsc is returned by 0x15 leaf.

Reply:
OK, then we can go with leaf 0x15 first.